proto-lens-protoc/src/Data/ProtoLens/Setup.hs

```diff
@@ -209,10 +216,41 @@ generateSources root l files = do
                      [ InstalledPackageInfo.dataDir info </> protoLensImportsPrefix
                      | info <- collectDeps l
                      ]
-    generateProtosWithImports (root : importDirs) (autogenModulesDir l)
+    -- Generate .hs files into a temporary directory, then move them over
+    -- to the target (autogen) directory only if they are different from
+    -- what's already there. This way, we don't needlessly touch the generated
+    -- .hs files when nothing changes, and thus don't needlessly make GHC
+    -- recompile them (as it considers their modification times for that).
+    -- See https://github.com/google/proto-lens/issues/176
+    let tmpAutogenModulesDir = autogenModulesDir l ++ "-protoc-tmpOutDir"
+    -- Generate .hs files into temp dir.
+    generateProtosWithImports (root : importDirs) tmpAutogenModulesDir
                               -- Applying 'root </>' does nothing if the path is already
                               -- absolute.
                               (map (root </>) files)
+    -- Discover generated files.
+    -- `getDirectoryContentsRecursive` is lazy IO, so we `force` through
+    -- the list to make it strict hereinafter.
+    !generatedFiles <- force <$> getDirectoryContentsRecursive tmpAutogenModulesDir
+    -- Move files to autogen dir only if file contents are different.
+    forM_ generatedFiles $ \pathRelativeToTmpDir -> do
+        let sourcePath = tmpAutogenModulesDir </> pathRelativeToTmpDir
+        let targetPath = autogenModulesDir l </> pathRelativeToTmpDir
+        identical <- do
+            targetExists <- doesFileExist targetPath
```

Comment:
What do you think about comparing file sizes before doing the full read? If you don't think it's worth the trouble because the common case is the files are identical, maybe mention that in the comment?

Reply:
@blackgnezdo I could do that, but I'd have to do it using [`withFile path ReadMode hFileSize`](https://stackoverflow.com/questions/5620332/what-is-the-best-way-to-retrieve-the-size-of-a-file-in-haskell/5621159#5621159), or add a dependency on [`unix-compat`](https://hackage.haskell.org/package/unix-compat-0.5.0.1/docs/System-PosixCompat-Files.html#v:fileSize) to [have it work on Windows while still being done done with `stat()` on other platforms](https://www.reddit.com/r/haskell/comments/3l8msi/is_there_a_portable_way_to_get_file_size/cv45hgy/).

In general comparing file size to eschew a read for only `stat` and `mv` is a good idea, but as you say, it's an optimisation for the case where the file does change, and in that case we have to recompile which dwarves the cost of the read, so it's your call on whether you think it justifies the extra logic. I have no opinion on it, but will happily add it or otherwise add a comment in this regard.